### ISSUE TITLE
Support a configuration option to disable wrapping optional attributes in MAYBE

### ISSUE THREAD

**Issue by mattiLeBlanc** [dotansimha/graphql-code-generator#10203]
### Is your feature request related to a problem? Please describe.

I get these TS issue in my code which forces me to either use maybe in my function input definition but I don't want to use that. I just use `{ tvName?: string }`.

![image](https://github.com/user-attachments/assets/57dad306-021a-4904-b553-a671eb395367)

It would be great if I can generate the Schema without any InputMaybe and Maybe.
I use generate types in my NodeJS backend so I dont have to maintain a secondary type library and just tap into the Schema types. That works great.

But I dont want to get rid of the Maybes in my  code everywhere, it is quiet annoying.
Like I have a function
```

async function getSomething({ id, name}: {id: string, name?: string}) {
....
}

and I call that function somewhere else using a payload coming from my graphql query (I use AWS Amplify).

const handler = async (event: SomeInput): Promise<any> => {
   const { id, name } = event;

 await getSomething(id, name) ;
}
```

I will now get a type error on `getSomething` because name is `InputMaybe<Scalars['String']>;` and my TS code only expected a string or undefined.
So now I have to cast my `name` or use something like `getSomething(id, name: name || undefined)` which is ugly.
Or I have to start using Maybe in all my functions and I dont want to because it will couple my code to the Generator.

### Describe the solution you'd like

Add a configuration option to generate the types without Maybes.

So instead of :
```
export type ConnectTvInput = {
  code: Scalars['String'];
  facilityId?: InputMaybe<Scalars['String']>;
  mode: TvMode;
  residentId?: InputMaybe<Scalars['ID']>;
  tvName: Scalars['String'];
};
```
we get
```
export type ConnectTvInput = {
  code: Scalars['String'];
  facilityId?: Scalars['String'];
  mode: TvMode;
  residentId?: Scalars['ID'];
  tvName: Scalars['String'];
};
```

### Describe alternatives you've considered

Run a regex Replace over my generated code and try to manually remove InputMaybe and Maybe, but it may be a bit difficult with nested maybes in Arrays.

**Comment by eddeee888** [comment]
Hello, 

You could use [inputMaybeValue](https://the-guild.dev/graphql/codegen/plugins/typescript/typescript#inputmaybevalue) to change `InputMaybe`  to just `undefined` e.g.

```
generates: {
  'src/types.generated.ts': {
    plugins: ['typescript'],
    config: {
      inputMaybeValue: 'T | undefined'
    }
  }
}
```

Note that a `null` value is a valid case in that your resolvers can receive, so be careful when you make this change because you can only do falsy checks:

```ts
{
  yourResolver: (parent, { input } ) => { // your input can be `T | null | undefined`, but if you've changed it to `T | undefined`...
    // then, you won't be able to do this safely:
    if( input !== undefined ) {
      console.log(input); // input is `T | null` at runtime but your type is saying `T`, which is wrong
    }

    // instead, you can only do falsy checks like `if(!input)`
  }
}
```

**Comment by mattiLeBlanc** [comment]
@eddeee888 I will give it a try.
I am not afraid of `input: null`. If I don't have a input in a query, my query will just look like
```
query {
  yourResolver {
    id
  }
}
```
If there are arguments, they are either required not required.

**Comment by eddeee888** [comment]
Closing this as this seems to fix the issue. Please let me know if it doesn't. Thanks!

**Comment by mattiLeBlanc** [comment]
@eddeee888 

I am not sure what I am doing wrong but I still see:
```
export type UpsertOrganisationInput = {
  id?: InputMaybe<Scalars['ID']>;
  id2?: InputMaybe<Scalars['String']>;
  musicPooling: Scalars['Boolean'];
  name: Scalars['String'];
};
```
The input in the schema looks like:
```
input UpsertOrganisationInput {
  id: ID
  id2: String
  name: String!
  musicPooling: Boolean!
}
```

My config looks like:
```
import type { CodegenConfig } from '@graphql-codegen/cli'

const config: CodegenConfig = {
  schema: ['./dist/config/*.graphql', 'scalar AWSTimestamp'],
  documents: './src/config/queries.graphql',
  generates: {
    '../../libs/shared/models/graphql/admin-api.ts': {
      plugins: ['typescript', 'typescript-operations', 'typed-document-node'],
      config: {
        inputMaybeValue: 'T | undefined',
        maybeValue: 'T | undefined',
      }
    }
  },
  hooks: {
    afterAllFileWrite: ['eslint --fix']
  }
}
export default config
```
I am still seeing InputMaybe for id and id2.

**Comment by eddeee888** [comment]
Hi @mattiLeBlanc , `InputMaybe` will still be there because it is a utility type and we use it to simplify codegen output.
The difference is in its declaration:

```ts
type InputMaybe = T | undefined; // instead of `type InputMaybe = T | null |undefined;`
```

**Comment by mattiLeBlanc** [comment]
@eddeee888 
So I am still going to have a TS error:
```
Argument of type 'InputMaybe<string>' is not assignable to parameter of type 'string'.
  Type 'undefined' is not assignable to type 'string'.ts(2345)
const id: InputMaybe<string>
No quick fixes available
```
but I guess do see that a property is now `T | undefined` it makes is possible to feed it into a function that expects a string IF I check for undefined before hand. With the Null option it would still cause an error.

okay, thanks.

**Comment by mattiLeBlanc** [comment]
@eddeee888 
Actually, this `Maybe` is still causing trouble. For example I have this exported type:
```

export type MusicPoolingSettings = {
  __typename?: 'MusicPoolingSettings';
  archivedSubscriptions?: Maybe<Array<Maybe<MusicPoolSubscription>>>;
  migrated?: Maybe<Scalars['Boolean']>;
  size: Scalars['Int'];
  subscriptions?: Maybe<Array<Maybe<MusicPoolSubscription>>>;
};

```
then in my code  I destructure archivedSubscriptions:
```
const { archivedSubscriptions } = musicPoolingSettings;
```
this variable can possible be undefined.

so this code will make my compiler unhappy:
```
archivedSubscriptions.filter(val => val.expiryDateUTC)
// val can be undefined atm
```

When I use a default value  to catch the undefined,
```
const { archivedSubscriptions = [] } = musicPoolingSettings;
```
this is still causing an error because of the stupid Maybe in my type declaration.
I have to enforce a type or do some weird coding just to make TS happy, all because of this Maybe.

So may original question/request stands: Can we get an option to NOT have maybe's at all. Because I use this tool to generate TS types from my schema (which is great) but I don't use it to query graphql.
So I don't care about Null's or undefineds and maybe's. I deal with that in my code.

**Comment by eddeee888** [comment]
Hi @mattiLeBlanc , 

TLDR; Short answer is we will not support a config to remove `Maybe` wrapping types. But you can _dangerously_ set it to just `T` instead of `T | undefined`. Please read on for more details.

---

From the generated type, it looks like your schema is like this:

```graphql
type MusicPoolingSettings {
  archivedSubscriptions: [MusicPoolSubscription]
  // ... other fields
}
```

In GraphQL, `[MusicPoolSubscription]` means the server may return `archivedSubscriptions` as `null`, or an array. And each element is `null` or `MusicPoolSubscription`.

Therefore, `Maybe` is the correct semantic here, and you are getting type errors because TypeScript is suggesting that `archivedSubscriptions` or its elements can be null. If you want non-nullable array with non-nullable elements, you can change your schema type to `[MusicPoolSubscription!]!`. Relevant info in the [doc](https://graphql.org/learn/schema/#non-null)

If you don't want to use the non-null modifier, the alternative - as suggested in the TLDR; - you can dangerously set your maybe value to `T` which means the server can never return null for any field (which is odd, but I'll leave it to you to make the decision 🙂).

**Comment by mattiLeBlanc** [comment]
@eddeee888 

I guess I have to declare my property in my schema like this:
```
archivedSubscriptions: [MusicPoolSubscription!]
```
So `archivedSubscriptions` can be optional but if it IS an array, it should have `MusicPoolSubscription` items.
That way after I have validated that `archivedSubscriptions` is not undefined I can safely use it is child data.

Look in general I just use your tool to generate typescript voor my Appsync Lambda functions, so I dont have to duplicate models. I realise I have to be very particular in the graphql schema to not get to many TS complains about the Maybe Null etc.
Thanks for explaining.
